neurodsp/sim/periodic.py

```diff
@@ -179,6 +179,97 @@ def sim_bursty_oscillation(n_seconds, fs, freq, burst_def='prob', burst_params=N
     return sig
 
 
+@normalize
+def sim_variable_oscillation(fs, freqs, n_seconds='auto', cycle='sine', phase=0, **cycle_params):
+    """Simulate an oscillation that varies in frequency and cycle parameters.
+
+    Parameters
+    ----------
+    fs : float
+        Signal sampling rate, in Hz.
+    freqs : float or list
+        Oscillation frequencies.
+    n_seconds : 'auto' or float, optional, default: 'auto'
+        Simulation time, in seconds. If 'auto', the simulation time is based on `freqs` and the
+        length of `cycle_params`. If a float, the signal may be truncated or contain trailing zeros
+        if not exact.
+    cycle : {'sine', 'asine', 'sawtooth', 'gaussian', 'exp', '2exp'} or callable
+        Type of oscillation cycle to simulate.
+        See `sim_cycle` for details on cycle types and parameters.
+    phase : float or {'min', 'max'}, optional, default: 0
+        If non-zero, applies a phase shift to the oscillation by rotating the cycle.
+        If a float, the shift is defined as a relative proportion of cycle, between [0, 1].
+        If 'min' or 'max', the cycle is shifted to start at it's minima or maxima.
+    **cycle_params
+        Parameter floats or variable lists for each cycle.
+
+    Returns
+    -------
+    sig : 1d array
+        Simulated bursty oscillation.
+
+    Examples
+    --------
+    >>> freqs = [ 5, 10, 15, 20]
+    >>> rdsyms= [.2, .4, .6, .8]
+    >>> sig = sim_variable_oscillation(1000, freqs, cycle='asine', rdsym=rdsyms)
+    """
+
+    # Ensure param lists are the same length
+    param_keys = cycle_params.keys()
+    param_values = list(cycle_params.values())
+
+    param_lengths = np.array([len(params) for params in param_values
+                              if isinstance(params, (list, np.ndarray))])
+
+    # Determine the number of cycles
+    if isinstance(freqs, (np.ndarray, list)):
+        n_cycles = len(freqs)
+    elif len(param_lengths) > 0:
+        n_cycles = param_lengths[0]
+    else:
+        n_cycles = 1
+
+    # Ensure freqs is iterable and an array
+    freqs = np.array([freqs] * n_cycles) if isinstance(freqs, (int, float)) else freqs
+    freqs = np.array(freqs) if not isinstance(freqs, np.ndarray) else freqs
+
+    # Ensure lengths of variable params are equal
+    if ~(param_lengths == len(freqs)).all():
+        raise ValueError('Length of cycle_params lists and freqs must be equal.')
+
+    # Ensure all kwargs params are iterable
+    for idx, param in enumerate(param_values):
+        if not isinstance(param, (list, np.ndarray)):
+            param_values[idx] = [param] * n_cycles
+
+    param_values = np.array(param_values).transpose()
+
+    # Collect params for each cycle separately
+    cycle_params = [dict(zip(param_keys, params)) for params in param_values]
+    cycle_params = [{}] * len(freqs) if len(cycle_params) == 0 else cycle_params
+
+    # Determine start/end indices
+    cyc_lens = [len(create_cycle_time((1 / freq), fs)) for freq in freqs]
```

Comment:
I don't think the `create_cycle_time` is strictly needed here - it's making whole arrays, when what we really need is the length. Within `create_cycle_time`, the length comes from `np.arange(n_seconds, fs)`, which ends up being equivalent to creating an range from 0 to `int(np.ceil(n_seconds, fs))` in integer steps. With this, we don't need to generate the array, as the length of this array is equivalent to `int(np.ceil(n_seconds, fs))`. 

Long story short - I think this line can be written as:
`cyc_lens = [int(np.ceil(1 / freq * fs)) for freq in freqs]`
This is a little more efficient, and saves an import. 

If you want to play with this - I basically checked that nothing diverges in this code block to convince myself:
```
n_seconds, fs = 0.234, 101
print(n_seconds * fs)
print(int(np.ceil(n_seconds * fs)))
print(len(create_cycle_time(n_seconds, fs)))
print(len(np.arange(n_seconds * fs)))
```